csharp/src/Google.Protobuf/WritingPrimitives.cs

```diff
@@ -0,0 +1,568 @@
+        /// </summary>
+        /// <remarks>
+        /// This method simply writes a rawint, but exists for clarity in calling code.
+        /// </remarks>
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteLength(ref Span<byte> buffer, ref WriterInternalState state, int length)
+        {
+            WriteRawVarint32(ref buffer, ref state, (uint)length);
+        }
+
+        #endregion
+
+        #region Writing primitives
+        /// <summary>
+        /// Writes a 32 bit value as a varint. The fast route is taken when
+        /// there's enough buffer space left to whizz through without checking
+        /// for each byte; otherwise, we resort to calling WriteRawByte each time.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawVarint32(ref Span<byte> buffer, ref WriterInternalState state, uint value)
+        {
+            // Optimize for the common case of a single byte value
+            if (value < 128 && state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)value;
+                return;
+            }
+
+            while (value > 127 && state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)((value & 0x7F) | 0x80);
+                value >>= 7;
+            }
+            while (value > 127)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)((value & 0x7F) | 0x80));
+                value >>= 7;
+            }
+            if (state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)value;
+            }
+            else
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawVarint64(ref Span<byte> buffer, ref WriterInternalState state, ulong value)
+        {
+            while (value > 127 && state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)((value & 0x7F) | 0x80);
+                value >>= 7;
+            }
+            while (value > 127)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)((value & 0x7F) | 0x80));
+                value >>= 7;
+            }
+            if (state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)value;
+            }
+            else
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawLittleEndian32(ref Span<byte> buffer, ref WriterInternalState state, uint value)
+        {
+            if (state.position + 4 > state.limit)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
```

Comment:
If there is not enough room available, instead of writing individual bytes and checking for the size for each byte, have you considered refreshing the buffer so that there is room available. For example:
 https://github.com/JamesNK/protobuf/blob/155774996be69135c05b30960e6ef016c530894c/csharp/src/Google.Protobuf/CodedOutputWriter.cs#L478-L489

Reply:
I can look into that. Initially I tried to be conservative and I wanted to preserve the CodedOutputStream behavior (which always finishes the buffer all the way to its end before refreshing) as much as possible.
Maybe this could be a followup PR?